File: 2017/app/src/main/kotlin/aoc/solutions/Day1.kt
package aoc.solutions

import aoc.utils.FileLoader

fun solveDayOne() {
    val input = FileLoader("day1.txt").readFile()
    println("Day 1 part 1: ${partOne(input)}")
    println("Day 1 part 2: ${partTwo(input)}")
}

private fun partOne(input: String): Int {
    var sum = 0
    for (i in input.indices) {
        if (input[i] == input[(i + 1) % input.length]) {
            sum += input[i].digitToInt()
        }
    }
    return sum
}

private fun partTwo(input: String): Int {
    var sum = 0
    val mid = input.length / 2
    for (i in input.indices) {
        if (input[i] == input[(i + mid) % input.length]) {
            sum += input[i].digitToInt()
        }
    }
    return sum
}

File: 2018/src/solutions/day_1.rs
use std::collections::HashSet;

use crate::utils::file_loader::FileLoader;

pub fn solve_day_one() {
    let input = FileLoader::new("./inputs/day1.txt".into()).read_lines();

    println!("Day 1 part 1: {}", part_one(input.clone()));
    println!("Day 1 part 2: {}", part_two(input))
}

fn part_one(input: Vec<String>) -> i32 {
    input
        .iter()
        .map(|s| {
            s.parse::<i32>()
                .unwrap_or_else(|_| panic!("Could not convert {} to int", s))
        })
        .sum()
}

fn part_two(input: Vec<String>) -> i32 {
    let mut seen = HashSet::new();
    let mut curr_freq = 0;
    for num in input
        .iter()
        .map(|s| {
            s.parse::<i32>()
                .unwrap_or_else(|_| panic!("Could not convert {} to int", s))
        })
        .cycle()
    {
        curr_freq += num;
        if seen.contains(&curr_freq) {
            return curr_freq;
        }

        seen.insert(curr_freq);
    }

    -1
}
